docs/DesignDocs/Protected_Code_Loader_Proposal.md

```diff
@@ -0,0 +1,161 @@
+typedef struct pcl_table_t_
+{
+	/* Current state of PCL: initailized to PCL_PLAIN */
+    pcl_status_e pcl_state;                   
+    uint32_t     reserved1[3];                /* Must be 0 */
+	// GUID must match GUID in Sealed blob
+    uint8_t      pcl_guid[SGX_PCL_GUID_SIZE]; 
+    size_t       sealed_blob_size;            
+    uint32_t     reserved2[2];                /* Must be 0 */
+	/* For security, sealed blob is copied into enclave */
+    uint8_t      sealed_blob[PCL_SEALED_BLOB_SIZE];
+	/* SHA256 digest of decryption key */
+    uint8_t      decryption_key_hash[SGX_SHA256_HASH_SIZE];
+	/* Number of RVAs */
+    uint32_t     num_rvas;                    
+    uint32_t     reserved3[3];                /* Must be 0 */
+	/* Array of rva_size_tag_iv_t */
+    rva_size_tag_iv_t rvas_sizes_tags_ivs[PCL_MAX_NUM_ENCRYPTED_SECTIONS]; 
+}pcl_table_t;
+
+### Sections left plaintext
+1. ELF header  - binary header
+2. Sections table
+3. Segments table
+4. Sections' names string table pointed by e_shstrndx(e.g. .shstrtab)
+5. .oeinfo section holds enclave's metadata(properties)
+6. .bss and .tbss
+7. sections required to construct dyn_info (.dynamic)
+8. sections holds the content provided by entries with index DT_SYMTAB, DT_STRTAB and DT_REL in
+   dyn_info (e.g. .dynsym, .dynstr, .rela.dyn)
+9. sections containing PCL code and data:
+   a. section ".pcltbl"  // Designated section for PCL table
+   b. .nipx, .nipd, .niprod, .nipd_rel, .nipd_rel_ro_local
+10. sections for debugging
+   .comment, .debug_abbrev, .debug_aranges, .debug_info, .debug_line, .debug_lc, .debug_ranges,
+   .debug_str
+
+## Elf sections to be encrypted
+Sections not mentioned above are sections containing IP information that need to be protected.
+Mainly those sections are:
+1. Code sections -- .text in elf file
+2. Data sections -- .data in elf file (initialized local/global variables)
+3. Read-only Data sections -- .rodata in elf file (const variables)
+4. sections containing relocation info related with the above items
+
+## Sealing/Unsealing the decryption key
+To deliver the decryption key in a secure way, the key used to encrypt the enclave should get sealed
+by key policy OE_SEAL_POLICY_PRODUCT on deliverting to Protected Code Loader for decryption in a 
+sealed_blob. The protected code loader will unseal this sealed_blob to get the decryption key. 
+
+## new APIs and libs
+### Encryption tool
+new tool for ELF image encryption - oeencrypt: placed in OESDK installation folder as the bin files
+new lib for section ".pcltbl" and decryption -- liboepcl.a: placed in OESDK installation folderd
+as the lib files
+
+### Modifications to OE SDK runtime lib
+No new APIs exposed to user.
+1. A new setting is defined in include/openenclave/host.h as the argument to support
+protected code loader enclave loading in API oe_create_enclave.
+typedef enum _oe_enclave_setting_type
+{
+	// PCL setting type
+	OE_ENCLAVE_SETTING_PCL = 0xac120002,
```

Comment:
I think my primary concern with this becoming part of the OE SDK public API surface is that it enshrines the existing PCL implementation as the standard model for key handling for PCL, which I don't think generalizes well. 

For example, the idea that the dev is responsible for passing to `oe_create_enclave` the sealed blob assumes that a different TEE can and should do the same, but it's not clear that handling the sealed key itself is ideal or necessary. Consider a strawman example instead, where instead of passing the blob, you pass a path/configuration to get the key in whatever form makes sense for the TEE; that provides more flexibility to generalize. For example, it could just be a file path to the sealed key blob for SGX, but it might be a device path to a trusted key store, or a network endpoint to fetch a key from.